### PR TITLE
Page behavior

### PR DESCRIPTION
**Pull request by BogdanScherban**
What has been done:
1) Page behavior is similar to Showcase;
2) All common code was moved to template files and imported in the plugins;
3) Custom toolbars;
4) Custom buttons and custom pagination.

Results on the video:
https://drive.google.com/file/d/1wCkDORP57toCFRL0UDZgwl355xifeqFc/view?usp=sharing

### CODE REVIEW

**Comment by BogdanScherban** [comment]
@tony-shannon @fzaninotto 

I remember that I promised try don't create large PRs. There are 48 files.

But I can explain the reason. We have 6 plugins, and there are four files in each plugin. 

So, 24 files there are plugin files
5 files - for custom button
5 files - for templates (edit/create/list/show/details)
3 files - for custom toolbars

Besides this, 5 files were removed.

File: src/core/common/Buttons/CancelButton.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+};
+
+/**
+ * This component returns Cancel button
+ *
+ * @author Bogdan Shcherban <bsc@piogroup.net>
+ * @param {shape} classes
+ * @param {func}  onClickFunction
+ */
+const CancelButton = ({ classes, onClickFunction }) => (
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CancelButton.js]: Function was renamed to "redirectTo"

File: src/core/common/Buttons/CreateButton.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
+import React from "react";
+
+import { withStyles } from '@material-ui/core/styles';
+import IconButton from '@material-ui/core/IconButton';
+import AddIcon from '@material-ui/icons/Add';
+
+import { MAIN_COLOR } from "../../config/styles";
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CreateButton.js]: I created theme by createMuiTheme(). This settings JSON can be very large, but I can optimize it later.

File: src/core/common/Buttons/CreateButton.js [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    }
+};
+
+/**
+ * This component returns Edit button
+ *
+ * @author Bogdan Shcherban <bsc@piogroup.net>
+ * @param {shape}  classes
+ * @param {shape}  history
+ * @param {string} createPath
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CreateButton.js]: Parameter was renamed

File: src/core/common/Buttons/CustomPaginator.js [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        display: "flex",
+        flexDirection: "row",
+    },
+    button: {
+        display: "block",
+        border: "1px solid #e5e5e5",
+        height: "100%",
+        width: "35px !important",
+        borderRadius: 0,
+        color: MAIN_COLOR,
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CustomPaginator.js]: Done

File: src/core/common/Buttons/CustomPaginator.js [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+     * @author Bogdan Shcherban <bsc@piogroup.net>
+     * @param {number} page
+     */
+    goToPage = page => {
+        this.setState(
+            { page: page },
+            () => {
+                const { resourceUrl, history } = this.props;
+                const { page} = this.state;
+                history.push("/" + resourceUrl + "?page=" + page + "&perPage=" + ITEMS_PER_PAGE)
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CustomPaginator.js]: Done

File: src/core/config/styles.js [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { MAIN_THEME_COLOR } from "../../version/config/theme.config";
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/config/styles.js]: Done

File: src/core/common/ResourseTemplates/ShowTemplate.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+import EditButton from "../../common/Buttons/EditButton";
+
+const styles = {
+    showBlock: {
+        width: '100%',
+        backgroundColor: "white",
+        margin: "15px 15px 15px 0px",
+    },
+    expansionPanel: {
+        height: "49px !important",
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/ResourseTemplates/ShowTemplate.js]: Unfortunately, it doesn't work without "!important"

File: src/core/common/ResourseTemplates/ListTemplate.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+        '& tbody tr:hover': {
+            backgroundColor: MAIN_COLOR + " !important",
+            color: "white !important"
+        }
+    }
+};
+
+/**
+ * This component returns template for List page
+ *
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/ResourseTemplates/ListTemplate.js]: Added

File: src/core/common/ResourseTemplates/EditTemplate.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+            paddingTop: "0px !important",
+            paddingLeft: "10px !important",
+            paddingRight: "10px !important",
+        },
+    },
+};
+
+/**
+ * This component returns block with template for plugin edit form
+ *
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/ResourseTemplates/EditTemplate.js]: Added

File: src/core/common/ResourseTemplates/DetailsTemplate.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+ */
+class DetailsTemplate extends Component {
+
+    state = {
+        view: this.props.mode,
+    };
+
+    changeViewType = viewType => {
+        this.setState({
+            view: viewType,
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/ResourseTemplates/DetailsTemplate.js]: Renamed

File: src/core/common/ResourseTemplates/CreateTemplate.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            paddingTop: "0px !important",
+            paddingLeft: "10px !important",
+            paddingRight: "10px !important",
+        },
+    },
+};
+
+/**
+ * This component returns common template for plugin Create form
+ *
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/ResourseTemplates/CreateTemplate.js]: Done

File: src/core/common/Buttons/CustomPaginator.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+        '&:hover': {
+            color: "white",
+            backgroundColor: MAIN_COLOR
+        }
+    },
+    activeButton: {
+        display: "block",
+        border: "1px solid #e5e5e5",
+        height: "100%",
+        width: "35px !important",
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CustomPaginator.js]: Done. This styles work without "!important"

File: src/core/common/Buttons/CancelButton.js [hunk 13]
```diff
@@ -0,0 +1,39 @@
+import React from "react";
+
+import { withStyles } from '@material-ui/core/styles';
+import IconButton from '@material-ui/core/IconButton';
+import BlockIcon from '@material-ui/icons/Block';
+
+const styles = {
+    button: {
+        display: "block",
+        width: "100px",
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CancelButton.js]: Fixed

File: src/core/common/Buttons/CustomPaginator.js [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            }
+        );
+    };
+
+    render() {
+        const { classes, rowsNumber } = this.props;
+        const { page } = this.state;
+        const buttonsNumber = Math.ceil(rowsNumber / ITEMS_PER_PAGE);
+        let buttons = [];
+        for (let i = 0; i < buttonsNumber; i++) {
```
[the remark this answers is not in the document]
Comment (BogdanScherban) [reply src/core/common/Buttons/CustomPaginator.js]: @fzaninotto @tony-shannon 

I added a space to digit buttons for case, when buttons number is more than maximal:
image

Maximal number is equal 5 (as constant).

**Comment by BogdanScherban** [comment]
@fzaninotto 

Thanks, Francois! I did all required changes according to your notes.

But only one I didn't change - about "!important" for block title. It doesn't work without it.

**Comment by fzaninotto** [comment]
Great, thanks! As for my review comments, instead of replying to each one, I advise you to hit the "resolve conversation" when you think your changes answer my remarks.

@tony-shannon, we're good to merge.

**Comment by BogdanScherban** [comment]
@fzaninotto Thanks! I'll do it)

**Pull request merged**